daemon/node_monitor.go

```diff
@@ -0,0 +1,151 @@
+	"encoding/binary"
+	"net"
+	"os"
+	"sync"
+
+	"github.com/cilium/cilium/daemon/defaults"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/byteorder"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+func (d *Daemon) receiveEvent(m *bpf.PerfEventSample, c int) {
+	pl := bpf.MonitorPayload{Data: m.DataCopy(), CPU: c, Lost: 0, Type: bpf.EventSample}
+	d.send(pl)
+}
+
+func (d *Daemon) lostEvent(m *bpf.PerfEventLost, c int) {
+	pl := bpf.MonitorPayload{Data: []byte{}, CPU: c, Lost: m.Lost, Type: bpf.RecordLost}
+	d.send(pl)
+}
+
+var (
+	isMonitoring = false
+	listeners    = []net.Conn{}
+	mutex        sync.Mutex
+)
+
+// handleMonitorConn handles all the incoming monitors.
+func (d *Daemon) handleMonitorConn(server net.Listener) {
+	for {
+		conn, err := server.Accept()
+		if err != nil {
+			log.Print(err)
+			continue
+		}
+		log.Info("Monitor connected")
+		mutex.Lock()
+		listeners = append(listeners, conn)
+
+		if !isMonitoring {
+			d.runMonitor()
+		}
+		mutex.Unlock()
+	}
+}
+
+// runMonitor starts monitoring.
+func (d *Daemon) runMonitor() {
+	if isMonitoring {
+		log.Fatalf("Monitor is already running")
+	}
+	log.Info("Starting monitoring traffic")
+	isMonitoring = true
+	c := bpf.DefaultPerfEventConfig()
+	c.NumPages = d.conf.MonitorNumPages
+
+	me, err := bpf.NewPerCpuEvents(c)
+	if err != nil {
+		isMonitoring = false
+		log.Errorf("Error while starting monitor: %s", err)
+		return
+	}
+	d.monitorEvents = me
+
+	go func() {
+		for isMonitoring { // TODO: handle listener disappeared
+			todo, err := d.monitorEvents.Poll(5000)
+			if err != nil {
+				log.Errorf("Error Poll %s", err)
+			}
+			if todo > 0 {
+				if err := d.monitorEvents.ReadAll(d.receiveEvent, d.lostEvent); err != nil {
+					log.Warningf("Error received while reading from perf buffer: %s", err)
+				}
+			}
+
+			// Configuration change restart
+			if d.conf.MonitorNumPages != d.monitorEvents.Npages {
+				d.stopMonitor()
+				d.runMonitor()
+				return
+			}
+		}
+	}()
+}
+
+func (d *Daemon) stopMonitor() {
+	log.Info("Stopping monitor...")
+	isMonitoring = false // TODO: do we need to lock this?
+	if d.monitorEvents == nil {
+		log.Info("monitorEvents is nil??")
+		return
+	}
+	d.monitorEvents.CloseAll()
+}
+
+func (d *Daemon) send(payload bpf.MonitorPayload) {
+	mutex.Lock()
+	defer mutex.Unlock()
+	if isMonitoring && len(listeners) == 0 {
+		d.stopMonitor()
+		return
+	}
+
+	buf := payload.Encode()
+	meta := &bpf.MonitorPayloadMeta{Size: uint32(len(buf))}
+	i := 0 // Used for removing lost clients.
+	for _, client := range listeners {
+		if err := binary.Write(client, byteorder.Native, meta); err != nil {
+			log.Infof("Removing lost client %s", err)
+			client.Close()
+			continue
+		}
+
+		if _, err := client.Write(buf); err != nil {
+			log.Warnf("Write: %s", err)
+		}
+		listeners[i] = client
+		i++
+	}
+	listeners = listeners[:i]
+}
+
+// EnableMonitor is used by the daemon to turn on the traffic monitoring.
+func (d *Daemon) EnableMonitor() {
```

Comment:
So the exec package is for running external commands. Do you want this to be a separate binary that is executed?

If that's not the case then I would continue with moving this to the  monitor package and look into keeping a reference to a new structure `NodeMonitor` to remove most of the daemon references. The daemon would configure this struct and instead of using fatal, return errors  and handle them.